File: backend/app/routers/goals.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session
from typing import List
from uuid import UUID

from .. import schemas, models
from ..database import get_db
from ..auth import get_current_user

router = APIRouter(
    prefix="/goals",
    tags=["Goals"]
)

@router.post("/", response_model=schemas.GoalResponse, status_code=status.HTTP_201_CREATED)
def create_goal(goal: schemas.GoalCreate, db: Session = Depends(get_db), current_user: models.User = Depends(get_current_user)):
    new_goal = models.Goal(**goal.model_dump(), user_id=current_user.id)
    db.add(new_goal)
    db.commit()
    db.refresh(new_goal)
    return new_goal

@router.get("/", response_model=List[schemas.GoalResponse])
def get_goals(db: Session = Depends(get_db), current_user: models.User = Depends(get_current_user)):
    return db.query(models.Goal).filter(models.Goal.user_id == current_user.id).all()

@router.put("/{goal_id}", response_model=schemas.GoalResponse)
def update_goal(goal_id: UUID, goal_update: schemas.GoalUpdate, db: Session = Depends(get_db), current_user: models.User = Depends(get_current_user)):
    db_goal = db.query(models.Goal).filter(models.Goal.id == goal_id, models.Goal.user_id == current_user.id).first()
    if not db_goal:
        raise HTTPException(status_code=404, detail="Goal not found")
        
    update_data = goal_update.model_dump(exclude_unset=True)
    for key, value in update_data.items():
        setattr(db_goal, key, value)
        
    db.commit()
    db.refresh(db_goal)
    return db_goal

@router.delete("/{goal_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_goal(goal_id: UUID, db: Session = Depends(get_db), current_user: models.User = Depends(get_current_user)):
    db_goal = db.query(models.Goal).filter(models.Goal.id == goal_id, models.Goal.user_id == current_user.id).first()
    if not db_goal:
        raise HTTPException(status_code=404, detail="Goal not found")
        
    db.delete(db_goal)
    db.commit()
    return None

File: frontend/lib/api/goals.ts
import { apiClient } from './client';

export interface Goal {
  id: string;
  title: string;
  description: string;
  category: string;
  priority: number;
  deadline: string;
}

export const goalsApi = {
  getGoals: async () => {
    const response = await apiClient.get<Goal[]>('/goals/');
    return response.data;
  },
  
  createGoal: async (data: Partial<Goal>) => {
    const response = await apiClient.post<Goal>('/goals/', data);
    return response.data;
  },
  
  updateGoal: async (id: string, data: Partial<Goal>) => {
    const response = await apiClient.put<Goal>(`/goals/${id}`, data);
    return response.data;
  },
  
  deleteGoal: async (id: string) => {
    await apiClient.delete(`/goals/${id}`);
  }
};
